### ISSUE TITLE
Build Communications

### ISSUE THREAD

**Issue by l-arnold** [l-arnold/tkl-nomadic-odoo#11]
I propose we move some of these conversations off of TRACKER and into Issues for tkl-nomadic-odoo as I am doing here.

Probably want to have @DocCyblade, @l-arnold and @JedMeister  watching (yep) - updated.  Also needed to bring JedMeister in as a Collaborator to get an @ to work.

**Comment by JedMeister** [comment]
Ah you meant @jedmeister! :smile:

**Comment by DocCyblade** [comment]
Sounds good. We can move the conversation here then

**Comment by DocCyblade** [comment]
Made some changes and doing a build now. This will be the last one until after 6 My time

**Comment by l-arnold** [comment]
Sounds good.  
Doing a build now as well (dev-wip1).

I think you are in the East (EST).  I am in Colorado (MST).  Need to get on some other projects for sure.  Would love to have this close to wrapped.

I put a few "issues" up we might address.  (the ARCH/REL/VER) thing I will take a stab at in dependencies.  We need it for Gdata and WHTMLTOX

**Comment by l-arnold** [comment]
How to make a Smile?   :)  I am guessing, but no. :smile: (colon : sm) and starts to show. OK
Thumbs made with colon :+1:

**Comment by l-arnold** [comment]
@JedMeister   If I "fetch" Common that is temporary.  Correct?  Doing a Pull is more permanent.
Will the fetch effect the next build as I assume?

**Comment by DocCyblade** [comment]
Fetch will pull the change down to your repo, pull will fetch and merge

**Comment by l-arnold** [comment]
(Updated x 2)
Build with Common "Fetch" to turnkey/fab/common 
USING dev-wip1 for build

Firstboot
Postgresql was not running at the end of first boot

2nd Boot.
Postgresql running
openerp can log in with mcookie assigned to /opt/openerp/odoo/openerp-server.conf

postgres can NOT log in with password set at firstboot inithooks (Was able to by assigning same pw in Webmin Postgresql USER update (postgres)

so exact same as yesterday.  Perhap the Common Fetch did not work

Update #1 Very Long Wait for Odoo.
It seems that the OpenERP Database could be accessed at login time (I bypass to manage databases) and create a new one.  Will see if I can get into OpenERP by itself.

Update #2  Openerp Database is "accessible" to Odoo.  However, I do not know the User Login or the Password for it.  Normally this would be set in "database creation" but perhaps there is a setting some where defining such.

(Some Merit in having an accessible database at start, however..  need to set controls to it as well)  I will look in Postgre to see if there is a user in there.  There must e an admin.  Perhaps the PW is setting there as well.

**Comment by l-arnold** [comment]
As I see you are working to "not" create openerp database per say, I think that is a good direction.
If we do create it we need to seed it with a user name and password and that seems complicated and un-needed right now.  

Reason for doing so would be to put in TKL logos, branding things like that, but that would be overwritten right away anyway I feel.

Just did a pull trying to have VER/REL/ARCh in 10-dependencies
Common next

Will report with a new build soon.

**Comment by l-arnold** [comment]
Not sure why I can't use
make clean && make 
any more.
Keeps telling me there is no rule

now tried
  make clean
then
  make

(seems to be building, but I hope it is correct)

what am I missing here?

**Comment by DocCyblade** [comment]
Just did a successful build with everything seems to be working. Going to do another build to see if this is working again and not just a fluke.

**Comment by l-arnold** [comment]
Cool.  I'm was prepping such too.  Seemes wierd in that I know you made some changes.
Just did "git pull" on your branch into the directory I used yesterday.  kens-dev-wip1
read this transcript and tell me if it looks ok.  Nothing updated which is the wierd part.
Same on the Common Branch just now:

root@tkldev products/tkl-odoo-dev# dir
kens-dev-wip1  tkl-nomadic
root@tkldev products/tkl-odoo-dev# cd kens-dev-wip1
/turnkey/fab/products/tkl-odoo-dev/kens-dev-wip1
root@tkldev tkl-odoo-dev/kens-dev-wip1# git pull
remote: Counting objects: 16, done.
remote: Compressing objects: 100% (16/16), done.
remote: Total 16 (delta 6), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (16/16), done.
From https://github.com/l-arnold/tkl-nomadic-odoo
   c8fd28c..737f613  dev-wip1   -> origin/dev-wip1
Already up-to-date.
root@tkldev tkl-odoo-dev/kens-dev-wip1# git branch
- kens-dev-wip1

**Comment by DocCyblade** [comment]
It seems that everything works fine, however only after a reboot or a restart of the Postgres and odoo services. 

Looking at doing a init hook to restart the services 

By the way, the default odoo login is admin admin

**Comment by l-arnold** [comment]
Cool.  Is that set in 10-main  ?  I did not see where you set that?
Perhaps in the "early version" openerp-server.conf

I am doing a build.  Last one failed because I needed to uncomment REL and ARCH in my whtmltox wget.  Lets see if it works now.

I am working in dev-wip1
can also build kens-dev-wip1

Should I do so before we try a Pull Request?

**Comment by DocCyblade** [comment]
My last build worked like a treat!

I pushed my changes check out my branch and do a clean build and test the ISO install

**Comment by DocCyblade** [comment]
I tried three times, each one built and ISO installed.

All services seemed to be working going to main website, asked to create a database!

**Comment by DocCyblade** [comment]
I am not sure you want to nest directories like that under products. I never have done that. 

I just create two directories odoo-1 and odoo2 under products.

**Comment by l-arnold** [comment]
Good idea.  I will get it flat.  Have several folders out under products is all.

my REV/VER/ARCH is not working right for wkhtmltox  (will revert) in 10-dependencies(dev-wip1)
(updated filename)

---- UPDATE
Just added GEN to separate from VER  (there is a folder with 0.12) that needs to be called before the full version call there.

will test again after kens-dev-wip1 which is running now  (unnested)

**Comment by l-arnold** [comment]
kens-dev-wip1  (my rc2 iso)
Just built,  installed clean

Postgresql running at firstboot
Got an error w/ Odoo at first boot  (Service not available)

Second Boot
Odoo came up after 3 x Refreshes  (not sure why the lag)
web/database/manager#action=database_manager

But Running.

I have a Snapshot saved.  Will revert to "restart point" and see if I wait at "First Boot" (post passwords) I can get Odoo to come up by its own.  May be I did not wait long enough.

Would be interesting to have a "console" that said "when ODOO is ready"  (complicated)

**Comment by l-arnold** [comment]
Round 2
Postgres Running
Logged into postgres via Adminer
Odoo Process "says" running  
   530  openerp 18:02   python /opt/openerp/odoo/openerp-server --config=/opt/openerp/odoo/openerp-serve ...

Odoo at First Boot Returns:
- Internal Server Error
- The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.

Odoo Second Boot
- again no stragglers
  *After a distracted wait on my side Odoo Boots fine
  *no database loaded so goes to (URL)web/database/manager#action=database_manager

Why don't we just put a Reboot in the InitHook?  Would this make sense?  I realize better to just reboot the services.

Looks great ALL THINGS considered.  THANK YOU.

**Comment by DocCyblade** [comment]
I am working off of all SSD drives. My wait time was like 10 sec. 

During first boot did you see it stop and restart the services?

**Comment by l-arnold** [comment]
Not visually anywhere.  Would that be in the install screen?  I will try again and see.  Goes quick because I have a snapshot to "reboot" that is part of the Install Protocol.

I will count to 10 and see whether I can get in after setting PW and taking in Updates (Updates could be part of the issue really - so I will skip them this next time)

**Comment by l-arnold** [comment]
No visual reboot of services at first boot.
Could have a "restart services" as part of the Advanced Menu...  That or some sort of "system check".
We could also just update
confconsole.txt  
and Say, Please choose "A" and Reboot your system One Time at First Installation  (not sure how we would handle a Amazon Install though.

(reordered EDIT  for clarity)

Still 500 Internal Server Error on First Boot

Just tried to "stop/start" openerp-server in "running services" in Webmin.  This time
503 Service Unavailable
The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.

Prior was 500 error not 503

(Moved line below)
Did need a second reboot to bring Odoo Up.  

Did Start Clean on Second Boot, no intervention.
(UPDATE)..  ACTUALLY STRAGGLERS show up on BLUE SCREEN.
A (advanced) B (back) clears them

**Comment by l-arnold** [comment]
maybe I don't have the "restart" code yet, but I think I do.
Also we are both running VMware version I believe  (not sure that matters really)

**Comment by l-arnold** [comment]
Updated Comment 2 above:
Did Start Clean on Second Boot, no intervention.
(UPDATE).. ACTUALLY STRAGGLERS show up on BLUE SCREEN.
A (advanced) B (back) clears them

**Comment by l-arnold** [comment]
in dev-wip1
Looks like I got my "ARCH" version of
wkhtmltox to download and dpkg.  Great for updates moving forward.  Complicated pull.

I have a build.  I won't install just now.  If its in its in.

Should we do a Pull Request into dev-wip1 ?
Should we do a Pull Request into dev ?

**Comment by DocCyblade** [comment]
What is your testing protocol?

What steps do you take from build to installing to trying to test the app? Maybe I am missing something in your steps

Mine is this

I have a VM that is blank with a snapshot to the blank
1. Make clean
2. Make
3. Copy the product.iso to my desktop
4. Start VM point at ISO
5. Use guided full, next next next
6. reboot 
7. Root password - 12345678
8. Postgres password - 1234567
9. admin password - 123456
10. Skip updates / email / tklbam
11. Shows that services are restarting
    12.check each port

**Comment by DocCyblade** [comment]
Just did another test and I had no issues.

**Comment by l-arnold** [comment]
I am basically doing the same except I don't see #11 anywhere.

In TKLDEV I do 
git pull
(occassionally say git branch to make sure I am where I want to be)
make clean
make
(wait a while)
In SFTP I rename the ISO using a convention like "kens-dev-wip-rc2.iso"
Download
Upload to my VM System
Define a new VM that boots from the ISO
Use guided Full, next, next, next
(make a machine snapshot Now before reboot)
reboot
root password (consistent)
postgres password (consistent)
admin password (consistent)  (ADDED to comment)
Skip TKLBAM, Skip Email, 
(Sometimes UPdate, sometimes Not)

If I do updates, perhaps it restarts all services then, however, no visual
(11 should be here, but I don't see anything.  MIght be doing it might not)

My system comes up to confconsole (blue screen)

Great that it came up.

**Comment by l-arnold** [comment]
Lets do a few Pulls (Pull Requests)
kens-dev-wip1 > into dev-wip1

test (this will merge with my ARCH update for 10-dependencie (whtmltox)
If good

dev-wip1 > dev

Thoughts?

**Comment by l-arnold** [comment]
You wrote:  11: Shows that services are restarting 12.check each port

(can you do a screenshot perhaps ?)

Might be that my old VMWARE doesnt' do that (Version 4.0 I believe)

**Comment by l-arnold** [comment]
I updated my last ISO to Sourceforge.  Maybe you can try it to see if you get the same result as your ISO:
https://sourceforge.net/projects/tklnomadicodoo/files/
called: kens-dev-wip1-rc2.iso

Some old ones there I should delete otherwise.  Will do a new GOOD one very soon.  Maybe we drop the Nomadic part in the naming.

**Comment by l-arnold** [comment]
@DocCyblade Not clear why I am not seeing more changes.  Maybe you did a merge already.
see this link:

https://github.com/l-arnold/tkl-nomadic-odoo/compare/kens-dev-wip1...dev-wip1

**Comment by l-arnold** [comment]
It is directional.  That comparison shows "my change" in dev-wip1

Reversing the order shows "your change"
https://github.com/l-arnold/tkl-nomadic-odoo/compare/dev-wip1...kens-dev-wip1

**Comment by l-arnold** [comment]
I can see from the Changes and My files that I do not have the "restart services" file.

overlay/usr/lib/inithooks/firstboot.d/97-odoo-restart-services

Does it need to be (forgot the word) chmod'd to be "executable"

**Comment by l-arnold** [comment]
Just ran and got the updates via PULL

root@tkldev products/kens-dev-wip1# git pull
remote: Counting objects: 26, done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 26 (delta 5), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (26/26), done.
From https://github.com/l-arnold/tkl-nomadic-odoo
   9a81b14..d6e41bf  kens-dev-wip1 -> origin/kens-dev-wip1
   737f613..4c36f15  dev-wip1   -> origin/dev-wip1
Updating 9a81b14..d6e41bf
Fast-forward
 conf.d/30_main                                      |  9 ++-------
 .../lib/inithooks/firstboot.d/20-odoo-db-secrets    |  1 +
 .../inithooks/firstboot.d/97-odoo-restart-services  | 19 +++++++++++++++++++
 3 files changed, 22 insertions(+), 7 deletions(-)
 create mode 100755 overlay/usr/lib/inithooks/firstboot.d/97-odoo-restart-services
root@tkldev products/kens-dev-wip1#

**Comment by DocCyblade** [comment]
Do a pull request, just don't approve it yet we can then see what's changing etc. I think I can approve them since I have access to the repo

**Comment by l-arnold** [comment]
Ok.  Working on a new iso build from kens-dev-wip1 also

PR next.

**Comment by l-arnold** [comment]
Which should be "base" ?
Which should be "compare"  ?

(ie, right and left) in https://github.com/l-arnold/  )

**Comment by l-arnold** [comment]
https://help.github.com/articles/using-pull-requests/

(left) = "base branch"  (normally goes to default branch .  I assume should be dev-wp1)
(right) = "head branch"  (I assume kens-dev-wip1)

**Comment by l-arnold** [comment]
Ran the PR.  Have a look if you would.

**Comment by DocCyblade** [comment]
Just did another build mine works out of the box.

Have not tried your branch yet

**Comment by l-arnold** [comment]
Getting there.  I am abou 50% on package builds.  Running make again.

Should all be good.

**Comment by l-arnold** [comment]
FULL START no reboot.

Postgres
Odoo
No Stragglers yet.

Nice One!

**Comment by l-arnold** [comment]
TKL Odoo RC1

What do you think?  Lets take a break for a while.
I think my little set of changes integrated as well.  Will test that to be sure but was mainly a parsing issue with variables in the download name.

**Comment by l-arnold** [comment]
How can we put a Marker in Right Now in the Build so we know where we are?

https://github.com/l-arnold/tkl-nomadic-odoo/commit/682a82d5a9c1dabe61e6063999750c2ff109c5ed
Uploading ISO TKL Odoo RC1 under Files at:  https://sourceforge.net/projects/tklnomadicodoo
(takes a while to get up there.  First one did not complete as I opened a new page)

**Comment by DocCyblade** [comment]
I can do a test on your branch and we can merge it into dev, and dev into master as a beta1

RC should only be when everything is done. There are still some things may need to be tweaked or bugs to be shaken out.

**Comment by l-arnold** [comment]
Do a Test on the Branch.  If looks good we can put it into dev.  Makes sense.

beta is a better term than RC for right here.

:+1:   Let me know your thoughts on system and also documentation.
I was putting some quick info in the documentation Link

https://odooIP:12324 will take you there  (4th Icon)

Should also fix up that landing Page getting the column aligned.

We will let Jeremy know when we go into Dev Branch.  (he is likely reading this being so short on projects right now :smile: )

**Comment by JedMeister** [comment]
Great work guys! :+1: 

Yeah @l-arnold I've got nothing much on at the moment...! :stuck_out_tongue_closed_eyes:

**Comment by DocCyblade** [comment]
@l-arnold I tagged this as beta v0.1 

going forward I suggest that if you find big issues or even small ons open an issue on your own repo here  to track it. Also lets others know there are known issues. I also created some issue tags for you so you can tag issues a release number like we do in TKL

**Comment by DocCyblade** [comment]
@l-arnold - I think your pretty much setup friend. I went ahead and tagged your issues. If you need help with something, you create an issue, tag it with a tag and release version and assigned it to me :smile: 

Oh, and as far as :beers: is concerned I like local micro brews :smile: If I ever get out to your part of the country I'll look ya up! 

I am about to start or should I say finish my two TKL projects RequestTracker and Invoice Ninja. I am also working on two other open source startups, one for an open source calendar/scheduler to visually track peoples schedules (going to use it at work) and an open source mass notification system, also hope to use at my place of work some day. 

 Thanks @JedMeister too for helping us

**Comment by JedMeister** [comment]
@DocCyblade - FWIW note that tags don't auto push or pull from/to a repo. You need to explicitly push/pull them... I forget the switch to include tags - perhaps `-a`/`--all`?

Thanks @DocCyblade for lending @l-arnold a hand. Got some great progress and found a really important bug that needed squashing in the process.

**Comment by DocCyblade** [comment]
@JedMeister - I had to look it up, I work on the command line, and I forgot what switch you used. the `git tag -a` will allow you to add notes to the tag. I think the tag shows up. I pushed it to "upstream" since my repo (fork from this one) is origin

```
git tag -a tag-id-here
git push --tags
```

**Comment by JedMeister** [comment]
Nice! :+1:

**Comment by l-arnold** [comment]
Thank you Ken @DocCyblade 
I may send you some beer over PayPal in the coming days..  
Definitely when you get out here lets get together.  

We both need a little rest on this one for sure.  Very cool it is building, secure and working as intended.  I will focus on some "use testing" next.  See what comes from that.

@JedMeister  What doe we need to do regarding TKLBAM?

**Comment by JedMeister** [comment]
Re TKLBAM here are the steps.
1. Make sure that the [core profile](https://github.com/turnkeylinux/tklbam-profiles/blob/master/core) includes all the places where Odoo has data/etc stored. (Remember that inclusions are just a path; exclusions are preceeded with a `-`).
2. Fork & clone the [TKLBAM profiles repo](https://github.com/turnkeylinux/tklbam-profiles); create a new local branch, etc etc.
3. Create a new tklbam profile (call it the appliance name e.g. `odoo`). Inside that file create a list of any additional paths to include or exclude (remember excludes preceeded by `-`). If there are none then you can make it blank. You'll notice that most of them are blank although TBH I think that almost all apps probably have some sort of cache that could be excluded.... Job for another day...

Note TKLBAM profile still won't work until the appliance is merged into the library and officially included...

**Comment by l-arnold** [comment]
Just a general process note:

*I think we are closing on our "Todo List" here for Odoo.  

*I have made a generally simple "doc" inside of the docs file that we may want to review for other "pointers".

*Samba can be removed after one cycle of PR's from dev-wip1 branch to dev.  We will do that by removing the first 2 lines of the Make File.
(on this note, could we keep those two lines but instead set it up so SAMBA is not starting unless done so from inside the Server ?  I think it does make a lot of business sense to have Samba be enabled for "close quarters" office work.  Not so sure if this gets loaded to Amazon.

*What work needs to be done to Prep this for "PreLoads" on Amazon etc?

**Comment by l-arnold** [comment]
Hi Guys.  I just did a Pull Request.  It looks like I am the only "participant".  Do you get notified?  Do you want to be notified.  Maybe just go look at it and give me a thumbs up.

I think it is all good safe changes.

**Comment by DocCyblade** [comment]
@l-arnold nice chat my friend! Was good to talk with you. I think we are on the same page. 

I did confirm that the mv command is not getting the .git folder. I think a better options would be to move the overlay files to the parent folder, so do does not exist and git clone directly into the openerp. Then move the conf files over. I am testing this now

**Comment by DocCyblade** [comment]
that seems to work better, you get to keep the .git

**Comment by DocCyblade** [comment]
Well bed time for me. I'l work on the .git issue. Need to re-arrange some code. I will change it once your done make any changes.

**Comment by JedMeister** [comment]
I assume that you are referring to [this mv](https://github.com/l-arnold/tkl-nomadic-odoo/blob/dev/conf.d/30_main#L72)? 

If so why not just replace [these lines](https://github.com/l-arnold/tkl-nomadic-odoo/blob/dev/conf.d/30_main#L66-L73) with this:

```
git clone https://github.com/odoo/odoo.git --branch 8.0 --depth=1  $ODOO_DIR
```

:question:

**Comment by DocCyblade** [comment]
Because of the config file already exists via an overlay.

We could just use the git hubs version of the config file and edit post git clone. I think it's a bare minimum and does not have all the comments. I will have to confirm this.

**Comment by JedMeister** [comment]
You could overlay the config file still but somewhere else then `mv` it in in the conf (instead of `mv`ing the whole git repo).

Another (perhaps nicer IMO) idea would be to overlay the config to ./etc (e.g. /etc/odoo/odoo.conf or whatever) and then symlink to it in the odoo dir (in the conf script). 

Actually IIRC it's the odoo initscript that applies the config to odoo (when it's started) so in that case you wouldn't even need to symlink - just adjust the path to point to the config file's (new) location (in /etc).

**Comment by l-arnold** [comment]
(EDIT FOR PHONE THUMBS) Glad we checked this.  I an sure we can work through one way or the other.

Any thoughts on how to update password Hash on TKLBAm?  (see tklbam issue)https://github.com/l-arnold/tkl-nomadic-odoo/issues/14

**Comment by DocCyblade** [comment]
Jeremy best idea, keep it in etc. don't need to symlink just need to update the init script to point to etc. makes since to keep config there anyway

**Comment by l-arnold** [comment]
(FIXED see next comment)   Just tested the "manual" call to the "Webconsole" 
- a2ensite odoo-12325.conf

For some reason after the last PR, 
CSS and JS folders from /var/www
and also most of the original "images" are not included
in /var/www/images

Able to Massage this back, but not understanding.  Perhaps the LAPP base of this is also changed.
I would like it to be functional with just a script call (a2ensite .. above)

**Comment by l-arnold** [comment]
Got it (~~)I think(~~).  We needED these two lines at the beginning of Makefile (from LAPP)
WERE REMOVED by LA last round  WEBCP!

COMMON_OVERLAYS = tkl-webcp
COMMON_CONF = tkl-webcp

Left Out
(~~)Not sure if we need the Lapp/main file to "remove unneeded icons"(~~)

(~~)*!/bin/sh -ex(~~)

(~~)\* remove unrequired webcp icons (all except those named) (~~)
(~~)\* cd /var/www/images(~~)
(~~)\* find . -type f ! -name "adminer.png" ! -name "shell.png" ! -name "tab.png" ! -name "webmin.png" -delete(~~)

**Comment by l-arnold** [comment]
dev-wip1  (ISO Build went well)
No Samba.  Several updates.
webconsole brought back in (should be easy now to remove, but I can't live without it as a Die Hard Webmin WebShell TKL services USER)!    (I think it is a very nice addition IMHO.)

**Comment by JedMeister** [comment]
I've lodged a PR for moving the init script and conf file into etc and some other tidying up...

**Comment by l-arnold** [comment]
How to Fix so we can Merge these Branches?

https://github.com/l-arnold/tkl-nomadic-odoo/compare/dev-wip1

The "dev-notes" in "Dev" are good, though ultimately will move into Docs or dissappear. I can find them if we delete.

**Comment by DocCyblade** [comment]
You want WIP into DEV? is this another beta release?

**Comment by l-arnold** [comment]
Yes, both woukd be great.

Your RST idea for apache is a good one.  Was looking for the same for odoo internal.may have found one yeatereday but need to come back to the subject.

Alternatively, we could put passwords on the doc system and have it function as doc/wiki.  KISS though is the direction now.

**Comment by l-arnold** [comment]
Moved From Final To Do list with a few edits

(EDIT)  THIS IS MOSTLY Commentary.  Will Move to "App Communications)
Quick Summary on my side.
It looks like we can get PostFix working by just including Some Documentation, otherwise we would need an Inithook to define the "catchall - email address" and set it into POSTFIX.  This is pretty easy and documented as to what should be set in the Postfix issue.

TKLBAM is effectively backing up and restoring via Shell.  The TKL Cert gets a little messy after wards.  Certified CERt Works fine.  All Data and Files move through very fast and without errors.  Again NOTE the TKL Cert is messy afterwards though.

(~~)WebConsole:  I still like it very much but have taken the a2ensite call for it out of the current Pull Request.  I have not heard anything otherwise so I will process the Pull so that I can also remove Samba.(~~)

This will leave the GIT Process Subject.  We really need this resolved to "assuredly" know that our DATA MIGATION issue is solved, but on my tests yesterday it was working as designed using the Process Bounce that pulled openerp.init start -u (database) -d (database) in the "UP BOUNCE" rather than the Service Call that stopped the service.

(~~)One Strong Rational for keeping the WEB CONSOLE in Place is that it provides an opportunity for better access to Documentation.(~~) Web Console deactivated: LA

I realize also, that perhaps the best "Documentation" foundation would be a WIKI on the Web rather than something internal.   A simple Link to that WIKI might be sufficient.

All that I can see on my side right now.  I think this is closing in.

**Comment by DocCyblade** [comment]
Hey guys, sorry for the sparse comments. Been very busy this weekend, and the next few days are booked got 15 hour days ahead of me. I am trying to keep up with the pulse of the conversation, Jeremy thanks for stepping in whew I left off, Landis sorry I have not been much help the last few.

Keep plugging away! I'll chime in when I can.

**Comment by l-arnold** [comment]
Give  that "dev" has not been building, for me at least,  I have change the "default" branch to 

dev-wip1.

Dev-wip1 is building well.  
Perhaps I can get Dev working,again but do not see time until early next week.

System is functionally working.  About 4 x "Documents can be moved to "dev-wip1".

Ideally the Dv branch can be brought to speed but will likely take some shell time.

I do feel we have a secur and functional build.  Can functionally be Beta 02.

Curious about your take on this.
best to bot!

**Comment by DocCyblade** [comment]
I'll have some time today to give it WIP a spin.

**Comment by l-arnold** [comment]
Thanks Doc.
Dev has a few docs I eant to move in but just can't build. Think i need a need a TklDev build.

If you can org / reorg branches that would be greay.  Should havr time FRI pm if any docs need to be moved or built.

Generally all rhe parts seem to be there.

**Comment by JedMeister** [comment]
Apologies on the radio silence. Been flat out with TurnKey and have been doing jury duty... Anyway FWIW dev builds fine for me, although I haven't tested beyond that. My guess is that there is something going on in your build environment that is upsetting things although it's really weird that the other branch builds ok for you...

**Comment by l-arnold** [comment]
Hi @jedmeister
9TYPO EDITS)

Sure would appreciate this being pulled into release 14.

I feel it has passed all the objective goals and been cleaned etc etc.

Put a lot of time into it too.

Saw it not listed today's announcement waving the flag on the subject.

Needed to fit some work in this week so couldn't be peppering the issue conversation here last few days.

**Comment by l-arnold** [comment]
Thumb typing from Nevada. Sorry for the typos.

**Comment by DocCyblade** [comment]
@l-arnold I was able to merge wip1 to dev no issues. I tagged it v0.2 and spun another branch dev-wip2 off of that version to make changes on if needed

@JedMeister I know you busy cranking it out, (good job by the way :+1: ) but if you can give us guidance to what needs to be changed before it's considered for 14.1 or a delayed 14 release. 

@l-arnold Once we know what needs to be changed if any, we can merge it back into dev then master and mark it v1.0 in master. This is what the TKL devs will use for the pull into an official repo. 

@JedMeister if there is anything I can do to speed this up, my weekend if free so I would like to jump on this.

**Comment by JedMeister** [comment]
@l-arnold - I'm really sorry but Odoo won't make it into v14.0. The final v14.0 images were built about a week ago. It's close but not quite close enough. I totally acknowledge all the hard work you've done but really v14.0 should have been released a month ago (if not before then) and if we kept waiting for one more thing it'd never get released.

We're not totally sure on how we will move forward but a big single release doesn't work all that well for anybody I don't think (end users maybe; but even then I'm not sure...). So we'll need to work that out... If we do a batch job of builds prior to v14.1 then Odoo will be in (assuming it's all good to go). Otherwise it will definitely be in v14.1.

So what is left to do?
I still haven't had time to have a really good (close) look over the code, but first thing that jumps to mind is that I don't think that any of the downloads are using the (TKLDev/polipo) proxy' that's not a massive thing but it should be resolved.

Also I am still not totally convinced that all users will want all of the included extras. TBH though I don't really know enough to be sure. An easy way to resolve that might be to whittle the actual appliance down to a MVP (minimum viable product) and move the extras to a post install script (so end users can decide if they want that. That would fit nicely with the intention for v14.1 to include extra/additional config and install options from confconsole. I'm not suggesting that would be the case for all of them (a quick google suggests that some would be universally useful but others I'm not so sure (e.g. the Magento connection stuff will only be useful for users who want to connect to Magento).

As for the conf script themselves I'd like to talk with Alon about them as they are arranged quite differently to the other appliances and keeping them consistent from my perspective is quite important. Although Alon may not be too concerned... Basically yours are organised in a way that makes sense when you are manually installing (e.g. download then install one bit; then download and install the next bit) whereas generally the conf in the other appliances batch all the similar processing together; (e.g. download everything that is needed in one script; install everything that is needed in anther, etc). There are some exceptions that use numbered conf scripts to separate parts but generally they are more complex than this install. However as I hinted above many of the additional bits (currently in the last script - and perhaps some of the first script) might be best put in a post install "helper" script that users can download and install themselves (if they want).

Also regarding the 2 dependencies in the first conf script: 
**gdata**
Firstly I'm still not totally convinced that gdata is needed at all (@l-arnold I know that you said it was definitely but what for?). And secondly (assuming you convince me it is required); where possible things should always be installed from package management (reduces maintenance and means it will get auto security updates). I accept that sometimes that won't work (not packaged or need a newer version, etc) but  I'm 99% sure that I already mentioned that`python-gdata` is the debian package name for this and I don't recall reading that that doesn't/didn't work.
**wkhtmltopdf**
Again this one is packaged. And again @l-arnold I recall you saying that the package didn't work; but I would be really interested in trying to discover why. My suspicion is that it wouldn't take much to make the packaged version work. And if there is something wrong with the Debian package then we probably should be reporting that to Debian. It should be the same version packaged in Jessie as what you are downloading and other than a difference in path of the installed scripts (deb pkg path is `/usr/bin` upstream pkg path is `/usr/local/bin`) and that the upstream package includes libraries and dependencies in it (whereas the deb pkg installs them separately) they seem very similar. I'd like to know more about this...

Both of those may seem small and insignificant but assuming that they are dependencies, then there may be longer term maintenance consequences of not using packaged versions; plus the obvious of not having security updates available... To be completely clear, I'm not saying don't use them; I'm just saying I'd need to see more rationale before I can accept them.

So @DocCyblade if you want to look at any of the above that might be useful?

Also I'd really like to make clear to both of you this is just my opinion. It is not set in stone. I am happy to defer to Alon if you'd like and/or happy for you to provide a rationale and/or argument in why I'm wrong and/or why what I say isn't relevant. However unless I've missed something important; at this point I stand by my opinion...

**Comment by DocCyblade** [comment]
@JedMeister Thank you for your honest input! 

@l-arnold I understand from talking with your goal in getting Odoo in the TKL arsenal for your own uses. I completely understand! On one side you want everything you are personally using, the other side TKL devs want it a different way to be included. This seems the two sides are far from each other but I think we can bring both of these together and get Odoo in the TKL library and in a form that will also be very useful to you as well.

I think the approach that should be taken is what Jeremy alluded to already. Do a base install of Odoo no extras or if you do the extras only do the most commonly used ones. (We can head over to the GITHub and pull stats to see what repos are hit the most). Switch out to use official deb packages where we can. IF there is a reason (especially if it breaks Odoo or the Odoo devs say to use it) then document that in an issue, and reference that issue in the conf script it's self. Try to align the build code with already TKL standards.

I think we can then bring this base install up to speed with some with some init-hooks that would ask what extra components they want. This will give the end use the power to install what they want at boot time, and maybe a confconsole screen to update that selection as well. This may take some time so, we could, to get the ball rolling get the base install done and get that add to the release, and work on a post install script that augments that base install for now until a nice clean way of doing this with init-scripts and confconsole can be done. 

@l-arnold if this sounds good to you, I'll pledge my support by doing all the heavy lifting of script editing. The only thing I don't know is the app it's self. If you would pledge to do testing of beta builds and file issues reports I think we could make a good team (I think we are already a good team!) but lets build on our strengths. You know Odoo and how it's configured as you use the software now, I am fairly good at VIM and TKLDev and GIT workflow. 

The cool thing about open source software is that, well its open. Your free to edit and use the TKLDev as you see fit for your own uses. So the GIT Repo is yours! So you could just use it as it. (I do that with some of my very custom build projects based on LAPP/LAMP) Pretty sure most if not all of the build scripts are covered by the GNU/v3 license. 

@l-arnold and @JedMeister What I purpose is this.
1. If Landis is good with the current beta, I'll start from there.
2. I will fork off of this current beta and clean up the branches but be sure to keep all git commit history from dev to show all the work that was already done.
3. Rename it tkl-odoo to keep separation of upstream (Landis copy) and the hopeful tkl included version
4. I'll create a few issues to keep track of things that Jeremy has said may need done.
5. I'll start work right away on things. 
6. Once I have successful build of an ISO, Landis can test the build himself and also install from ISO and report back any issues. May want to come up with a testing matrix we can use to test each release for bugs, to ensure the software is working like it should, and it's not something we left off. 
7. Once the work is done, tag it as ready.
8. Once the finally basic Odoo build scripts are complete, we can work on augmenting it will post install scripts that I will base mostly on your already there install scripts.
9. This would include, (exciting for me, learn python) build init hook to add modules, and a confconsole extension to add additional modules. 

Guys let me know you thoughts I am willing to do the fun(hard) work, I don't want to hijack Landis work either, just want to help out. I m busy most the day today, but this evening and early morning will be free. @l-arnold let me know if your good with the above and I get started tonight to early Sun AM.

**Comment by l-arnold** [comment]
Lost my post just now.

3 Things.
1. It needs to work
2. It needs to be secure
   3 It needs to be more than ISO
   4 It needs to allow customization and a systematic approach to tying in addons.

There has bern considerable yime and effort put into GData and Wkhtmltox.  Without. Wkhtmltox you cannot generate a fotmatted report or email and though I also started withvyhevwkhtlmyopdg package, it did not work.  Its not functioning with Odoo was widely reported.  

Scripting the pull eith version and generation calls is no different effectively than doing a Git Pull on Branch 8 on Odoo.

I do not recall the GData reason right now but again, it is widely required, I believe also by Odoo in their cookbook.

A diff Scripted nstall approach, invluding Run Tests on development builds is documented in OCA for Magento Connector.  I woukd take that to be an excellent point of reference.

I am travelling all day today.  I have a review in pricessbof OCA (Odoo Community) modules.  I will make a list of their purpose and general relavence.  "Mass Editor" , for instance, is a must have addition.

Tie ins to yhevOpen Source and Community aspect of Odoo that otherwiise gets obscured by Odoo SA should be one of the real goals of the aTKL Ap in my opinion.

So @DocSyBlade, I am in agreement with your proposal.  @JedMeistet also I am generally fine withvthings outlined.  Wevhave both posted on the forums about WKhtmltox and. Wkhymltopdf.  I will say that in the spirit of encouraging community involvement real discussions should be encouraged. This sometimes feels a bit like the Wizard of Oz where we can't see behind the screen.  

Referentially, I prob had a month of inithook crashes before Doc Came on board trying to set security where in retrospect I prob got it right a few times  but because the pgsqlconf.py script was broken I waas blaming myself and felt pretty close to abandoning the project.  We worked through that eventually.  There have been several big pushes rom both of you in this, butvwevhave been working from the goalbof "getting done" and into the library.

Methods and pulls I have put in were designed to be time and learning savers.  Pulling various several parts og the build will just make work for users ehen removed shoukd they want them. (the magento connector particularly).

That being said. I am perfectly capable of TKLBaming a Connector build over a generic build if need be.

We woukd definitely need a Branch to capture those customizations and I would like a way to move a customized branch forward with a generic branch.  I do not know how to do that approach however.

Anyway, please do "what needs to be done."  I expect this to be one of TKL''s most compelling builds as Odoo is not by itself,  all that easy to install.

Best to both of you,  i shouldbhave time again Sunday (tomorrow) or Monday.

Landis

**Comment by l-arnold** [comment]
Will edit above when I get laptop connected, hopefully today still.

**Comment by DocCyblade** [comment]
Sounds good. I'll get started tonight setting the structure up for the tkl-odoo repo. Look forward to get this done!

**Comment by JedMeister** [comment]
I'm just going to attack these one at a time.
**wkhtmltopdf**
@l-arnold I recalled you saying wkhtmltopdf doesn't work but I didn't realise that it not working had been documented. I did a quick google and I'm still not sure that it is an issue (not that I didn't believe you but I'd like more details so I can explain to Alon when he inevitably asks why it's not installed from the repos).

So I found quite a few references to wkhtmltopdf:
- [upstream install docs](https://www.odoo.com/documentation/8.0/setup/install.html) saying that Deb repo version doesn't work:- but my suspicion is that is referring to the Debian Wheezy (aka 7/TKL v13) version; not the Jessie (8/TKLv14) version
- [upstream forums](https://www.odoo.com/forum/help-1/question/openerp-v8-wkhtmltopdf-0-12-0-error-bug-52041) - This looks like it was written a while ago. It appears to be referring to a previous version of wkhtmltopdf (avaialble from the Ubuntu repos at the time) - not clear to me that it is still an issue.
- [upstream github issue](https://github.com/odoo/odoo/issues/5177) - this relates to a newer version of wkhtmltopdf not being compatible.

So from my end there is still no clear documentation that the Debian Jessie version won't work. But even if we did use wkhtmltopdf from wkhtmltopdf upstream (as you currently do) then perhaps we should be using a newer version (than 12.1)?

So IMO the way to go with this would be to install the repo version (`apt-get install wkhtmltopdf`) and reproduce the error (sounds like you print a report). Clearly document the issue (i.e. exact error messages on screen/ on the pdf / copy/paste from logs / etc). It'd also be great if steps to reproduce the error were included (I'd probably assume Odoo installed as your starting point - rather than include full install steps too)

Once you have documented the error (and assuming that you can reproduce it) then we need to consider the next step to try. So at this point I am assuming that the Jessie version is still an issue. Next step is to see if a newer version is available as a backport. And unfortunately there is no Jessie backport (look [here](https://packages.debian.org/jessie/wkhtmltopdf)). There is a newer version (0.12.2.4) in Stretch (Debian testing) but I think using packages from testing is probably a bad idea although not sure whether that would be worse than upstream or not. I'll need to check with Alon. Also we (as in TKL) should probably document this 'hierarchy' of where to pull software/packages from for the future reference of app builders...

**Comment by JedMeister** [comment]
**gdata**
Again I did some googling and it seems that gdata is a legitimate dependency (which still surprises me but I'm happy to accept it). However all the instructions I found that mention installing from upstream were all quite old (prior to Jessie release). The version that your conf currently installs is 2.0.17 but 2.0.18 is in the [repos](https://packages.debian.org/jessie/python-gdata). Unless there is some good reason (i.e. Odoo does not work with 2.0.18) then I'd use the repo version. If it doesn't work then document it as per above.

Also if we do use upstream then we should also be following the [FHS](http://www.pathname.com/fhs/). So the source code should really be in /usr/local/src I would think; although I guess in `/opt` is not that big a deal but should still probably be `/opt/src` or `/opt/odoo/src` (or similar).

**Comment by JedMeister** [comment]
**Additional modules**
I'm sure that some of them make tons of sense and I'm not suggesting that we include nothing extra but like I said stuff like Magento connection will only be useful for those that use both Odoo and Magento and so I don;t think it passes the "useful to most users" test...

As for installing the additional modules that aren't needed or universally useful; at this stage I would simply move them from conf scripts into a separate script(s) - perhaps in `/opt/odoo/bin` for now?

Like I think I mentioned for v14.1 we intend to extend the functionality of confconsole. Installation of additional modules would be an excellent candidate for that IMO. They could theoretically be included as an inithook but if that's done for it to be supported by headless deployments (e.g. Hub, LXC, OpenStack, etc) then they _need to have a default value_ (of "no") otherwise booting of headless deployments will hang (waiting for input but no way of giving that input). IIRC I already fixed the existing Odoo inithooks to support preseeding and headless deployments.

**Comment by JedMeister** [comment]
Also a couple more points:
FYI one of the next jobs on my list is to get the other build types (other than ISO and Hub) sorted out. So hopefully soon you will be able to build straight to VMDK/OVF. IIRC you have VMware don't you @l-arnold? Assuming you do then I imagine having an OVF will make the testing process much easier...

Anther thing I just noticed; the apache.conf sites could be condensed into one.

(And elaborating on points made earlier) to bring the conf scripts into line with other apps they could be rearranged into 2 conf scripts. A "download" (or "clone-repo"?) conf which downloads/clones Odoo, any upstream dependencies and required or generically useful Odoo modules - make sure it uses TKLDev's caching proxy. And a "main" conf script (which does all the install). Additional (required or generically useful) modules (cloned/downloaded in the first script) would be installed here too. Other optional modules could go in their own script (not using TKLDev caching proxy as that is only available at build time).

PS sorry to spam everyones inbox with a ton of comments; I just thought that it was better to try to split them up a little rather than have a single super long comment...

**Comment by DocCyblade** [comment]
Great feed back Jeremy!  

Landis, this all make sense to you? It does to me, I see where Jeremy is going with his logic and agree.

I think the best approach is to test the components that are in question gdata and the others. I'll create branches to test these specific issues. If you want to do some leg work to see what modules would be best to include in the base install start looking that up on the odoo forms see what's what.

**Comment by l-arnold** [comment]
I'm good with all.  Fine if packahes work, kust think I tried already.

Inithook choices would be fantastic.  

Will follow soon.
Utah desert sunset.

Best,
Landis

**Comment by JedMeister** [comment]
Re wkhtmltopdf
From what I can gather originally there was a need for a very specific version of wkhtmltopdf but that seems to be resolved now so any version from 0.12.1 on should work. Although the Debian repo version is 0.12.1 (and you already reported that that didn't work)...

Re gdata
I don't recall you mentioning that you tried the repo version but I may have missed it. My googling suggests that at one point upstream gdata was needed because the repo version was too old (in earlier debian/ubuntu versions) but the version in the Jessie repos is slightly newer than the one you are currently manually installing... Google also turned up Odoo 8 install instructions/tutorials showing installation of `python-gdata` with apt so my money is on that being a go-er...

**Comment by DocCyblade** [comment]
@JedMeister I'll try the build with the said packages instead, also when using GIT to pull down software in TKLDev is there a TKL way of using the proxy for GIT? I know I can change the proxy.http/proxy.https config, did not know if there is a better way. If there is already a good way of doing this point me to the source :-)

**Comment by JedMeister** [comment]
@DocCyblade - Sry for slow reply. TBH I never actually tested to see what works best; but according to my research git abides by the $http_proxy (and $https_proxy) system variables so something like this should do the trick:

```
[ "$FAB_HTTP_PROXY" ] &&  http_proxy="$FAB_HTTP_PROXY"
```

**Comment by l-arnold** [comment]
I am not clear on the purpose of using http_proxy for Git in this situation.  Is it not just for the build or is it for "post pulls". after ap install (ie in configurable inithooks)?

**Comment by l-arnold** [comment]
@DocCyblade Let me know when you want any testing done.  I am going to try to build a new TKLDEV.
@JedMeister Do you know if the full 14x release of TKLDEV is adding all components or do I need to also run the tkldev-setup as before?

**Comment by DocCyblade** [comment]
Landis, the proxy is only for the build process. Keeps you from requesting the GIT repos over and over (saves time and bandwidth) from building over and over

Will start working on a base install once I get the kids in bed. Should be soon

**Comment by JedMeister** [comment]
@l-arnold FWIW any TKL code that makes use of FAB/fab is almost certainly build related. Fab is one of the fundamental TKLDev tools. If you watch an appliance building you will see lots of lines that include `fab-apply-overlay`, `fab-install` and if you want to test a chroot (before building to ISO) then you can use `fab-chroot` to enter the chroot.

& the TKLDev v14.0RC2 and the stable release are essentially the same. The only thing that changed was the installer (`di-live`) so once installed and running they would be more-or-less identical IMO. A fresh install will run `tkldev-setup`. So if you have a fresh install then it should run as you install and be up to date. If you have an RC TKLDev build then run `tkldev-setup` to get it up to date (IIRC `tkldev-setup` was fixed to update if directories already exist between rc1 & rc2).

**Comment by DocCyblade** [comment]
@JedMeister FWIW: I found the below snippet of [code here](https://github.com/turnkeylinux/common/blob/master/conf/phpsh#L5) regarding GIT and TKLDev proxy.

```
[ "$FAB_HTTP_PROXY" ] && export HTTP_PROXY=$FAB_HTTP_PROXY
git clone https://github.com/facebook/phpsh.git $PHPSH
unset HTTP_PROXY
```

I don't see however any GITHUB cached by checking `ls -l /var/cache/polipo/`

```
root@tkldev-local products/tkl-odoo# ls -l /var/cache/polipo/
total 64
drwx------ 2 proxy proxy  4096 Sep 20 04:04 archive.turnkeylinux.org
drwx------ 2 proxy proxy  4096 Sep 20 04:06 debian.cc.lehigh.edu
drwx------ 2 proxy proxy  4096 Sep  8 22:43 debian.csail.mit.edu
drwx------ 2 proxy proxy  4096 Sep  8 22:42 debian.gtisc.gatech.edu
drwx------ 2 proxy proxy  4096 Sep 20 04:06 debian.mirror.constant.com
drwx------ 2 proxy proxy  4096 Sep  8 22:42 ftp-nyc.osuosl.org
drwx------ 2 proxy proxy 20480 Sep 20 04:06 http.debian.net
drwx------ 2 proxy proxy 12288 Sep 20 04:06 mirror.cc.columbia.edu
drwx------ 2 proxy proxy  4096 Sep 20 04:06 mirror.us.leaseweb.net
drwx------ 2 proxy proxy  4096 Sep 20 04:05 security.debian.org
```

**Comment by DocCyblade** [comment]
@JedMeister I also tried the following, does NOT seem to work. 

```
[ "$FAB_HTTP_PROXY" ] && git config --global http.proxy $FAB_HTTP_PROXY
git clone http://github.com/odoo/odoo.git --branch 8.0 --depth=1 $ODOO_DIR
git config --global --unset http.proxy
```

What did work was, from the command line:

```
root@tkldev-local fab/products# git config --global http.proxy http://127.0.0.1:8124
root@tkldev-local fab/products# git config -l
user.name=DocCyblade
user.email=ken@robinsonshome.net
http.proxy=http://127.0.0.1:8124
root@tkldev-local fab/products# git clone http://github.com/DocCyblade/mysql.git
Cloning into 'mysql'...
remote: Counting objects: 115, done.
remote: Total 115 (delta 0), reused 0 (delta 0), pack-reused 115
Receiving objects: 100% (115/115), 121.89 KiB | 0 bytes/s, done.
Resolving deltas: 100% (43/43), done.
Checking connectivity... done.
root@tkldev-local fab/products# ls -l /var/cache/polipo/
total 68
drwx------ 2 proxy proxy  4096 Sep 20 04:04 archive.turnkeylinux.org
drwx------ 2 proxy proxy  4096 Sep 20 04:06 debian.cc.lehigh.edu
drwx------ 2 proxy proxy  4096 Sep  8 22:43 debian.csail.mit.edu
drwx------ 2 proxy proxy  4096 Sep  8 22:42 debian.gtisc.gatech.edu
drwx------ 2 proxy proxy  4096 Sep 20 04:06 debian.mirror.constant.com
drwx------ 2 proxy proxy  4096 Sep  8 22:42 ftp-nyc.osuosl.org
drwx------ 2 proxy proxy  4096 Sep 20 04:56 github.com
drwx------ 2 proxy proxy 20480 Sep 20 04:06 http.debian.net
drwx------ 2 proxy proxy 12288 Sep 20 04:06 mirror.cc.columbia.edu
drwx------ 2 proxy proxy  4096 Sep 20 04:06 mirror.us.leaseweb.net
drwx------ 2 proxy proxy  4096 Sep 20 04:05 security.debian.org
```

Not sure why during the build its not working

I am heading to bed, will try some more in the AM or PM

**Comment by l-arnold** [comment]
Ok.  Start on comprehension.

It would be great if I could use FAB but build full ISO's.  I had understood I needed to go into CHROOT to do that (which I believe is how I broke my TKLDEV)....  Will do some studying for sure.

But back to the original subject.  Sounds like the code would call the Proxy and "once built" stop calling the full Git on sequential builds.  

I assume my "make clean"  then "make"  (or "make clean && make") will by pass the proxy?  Perhaps not.    Assume the purpose of this is to use the same Git Pull for every iteration of the App Build so likely to not require the "make clean" but something like "make cloud" "make (etc etc)".

All sounds good.  Need to study the docs.

**Comment by JedMeister** [comment]
@DocCyblade - great detective work!

I had read that "HTTP_PROXY" (all caps) doesn't work with git but that you need to use "http_proxy" or "https_proxy" (no caps) instead. 

As I said I never actually tested/checked (as I probably should have) but took what I had read and used it. 

To elaborate on my understanding: 
you would set "http_proxy" if using the git protocol e.g `git clone git://github.com/...`
or set "https_proxy" if cloning http(s) e.g. `git clone https://github.com/...` (AFAIK github http redirects to https)

However, seeing as you have found a technique that appears to work perhaps we can add that in the download conf script and remove it at the end? Or leave it there? Or something else? Also if we go that way then we probably should double check that it will work on a vanilla (unconfigured) TKLDev server (i.e. with the other bits of ~/.gitconf unconfigured - we could test by moving the fully configured file out and retry your steps above).

@l-arnold fab is a component of TKLDev. when you issue the `make` command TKLDev uses `fab` to build the ISO... Along the way you can use `fab-chroot` to test your build. E.g.

```
cd tkl-nomadic-odoo
make root.patched
fab-chroot build/root.patched
service apache2 start
service postgresql start
service openerp-server start
```

Then you should be able to browse to your TKLDev IP; port 80 (i.e. http) in your web browser and see Odoo web UI! When you are done, exit the chroot and kill all processes running in your root.patched chroot:

```
exit
fuser -k build/root.patched
```

After testing you can finish building with make:

```
make
```

Note though that any changes you made in the chroot will be lost. If you want to make changes they'll need to go in the overlay or conf scripts...

The proxy is purely for caching while you are building. Usage of the proxy applies only while working within TKLDev. It means that you don't need to keep downloading the same files over and over everytime you `make clean && make`.

**Comment by l-arnold** [comment]
Thank you for the instructions @JedMeister .

I will see what comes of the FAB (ulous) Proxy work you and @DocCyblade are working on and give it all a spin.

I will try the "fuser -k" command on my errant TKLDEV if it gives me troubles again as well.

**Comment by DocCyblade** [comment]
Nice pun!

**Comment by DocCyblade** [comment]
@JedMeister so I did some more testing, it seems it's not "caching" like we think it should. After reading (http://wildlyinaccurate.com/a-hackers-guide-to-git/) then my head exploding. I quickly picked up the bits of my head and figured the reason why. GIT is not pulling data that the proxy can understand and cache. 

So if we really want to "cache" GIT repos, we would need to create a custom script that does it for us. Not sure what the best way would be, but I think some script bash/python look something like this:
 `git-clone-proxy [options] [repo url] [git-clone options]`

The script would do the following 
1. Check to see if the repo has been requested before
2a. (No) Do a git clone in the cache directory
2b. (Yes) Go a GIT fetch to make sure it's updated. (Could set this on a time stamp to skip this)
3. Do a git clone from the cached clone passing all options given to the command line
4. Since we cloned it from file location, change origin to original url passed

Some possible options the script could implement
- Option to remove .git from final directory
- Force fetch/update

May want to base the above on this 
(http://randyfay.com/content/reference-cache-repositories-speed-clones-git-clone-reference)

Should I submit this as a feature request for FAB/TKLDev maybe a fab-git-clone or something that would keep cache? 

I am working a script now that will do this :-)

**Comment by JedMeister** [comment]
@DocCyblade - your awesomeness never ceases to amaze me! :+1: 

Sounds like a great idea all round (dev of the tool, lodging as a feature request, etc).

**Comment by DocCyblade** [comment]
@JedMeister I have good ideas every so often, just happens the moon and stars aligned I guess!

I will put in a feature request here soon. I started a shel script to do this. I used a template I found for shell scripting that I liked. I'll work on this as I have the time :-)
https://gist.github.com/DocCyblade

**Comment by l-arnold** [comment]
@DocCyblade and @JedMeister 
So, new build, here is my thought.
Inside /turnkey/fab  there could be a "git-cache" folder much like "common".
Could then go into the "git-cache" folder and do a "Git Clone" command as is done normally inside Products.

Having done that the git-cache (and its subfolders) could be addressed by TKL builds just as common is addressed.

Seems this functionality is actually already in place somehow as it is working with common.

**Comment by DocCyblade** [comment]
@l-arnold I am currently here (https://github.com/DocCyblade/tkl-odoo) working on building a basic install with no extras, the next thing to do would then be to test it. The idea is to start with a basic build, try and mimic as close as we can to the Odoo documentation, keeping service accounts and such in line with official documentation as well.

If you want to start making a test list we can use for each release. I have outlined the released in steps or milestones (https://github.com/DocCyblade/tkl-odoo/milestones) 

So here is how I think this can best work.
1. I will work on v0.3 a base build.
2. While I do this, if you can come with a test matrix we can use. Maybe a list of things to test in order we will test them.
3. Also if you can see what modules Odoo users are using, put them in a list.
4. Once I am done with v0.3, you can start testing it. I'll post the ISO somewhere you can access them
5. Ill open an issue to track the testing of the release. You can mark each test as done as you do them
6. While you start testing, I'll start working on v0.4
7. Any issues or bugs you find that it fails the tests matrix mark on the release issue and open a new issue.
8. Any issues will be address then pulled into the v0.4. 

We will release this build, release, test until we get what TKLDev find acceptable. I would like to open document testing and any decisions openly and in a systematic manner so that TKLDevs and Jeremy can see everything progress and why something was done the way it was etc.

I have added you to the repo so I can assign you issues. Just don't edit any code via git hub on this repo. If you want to make changes, create a new branch, make changed then pull request. 

Going to move future discussions here (https://github.com/DocCyblade/tkl-odoo/issues/1)

**Comment by l-arnold** [comment]
Sounds generally good Ken.  Would there not be sense in just making a "bare" branch in the current build (I do understand that is what is being built).

Just working on the "module" list but it is a very "big one" that is a bit daunting.  What would seem practical would be a way to "hook into the various repositories.  I think Odoo / addons by nature are the "default".  Ones like OCA are outside the Odoo repositories.  I will focus on the ones "outside".

 I don't know that we can pull off the "optional" Module install but I do feel there would be great value in that approach.  Could also apply to some other apps as well if a system were in place.  This is the part where Odoo, given their corporate druthers, will complicate things for un-educated users and where the "open" source starts to fall out of focus.

I would be curious if we could build some scripts that fire off the effect of each of the sections (and similar) in
conf.d/50_external_modules

I will test and see if I can keep up with the testing matrix.  Should not be a problem.  Will endeavor to mimic aspects here as well.

**Comment by JedMeister** [comment]
@DocCyblade - Love your work mate. IMO your process is one that I think all current and future TurnKey devs could benefit from (BTW IMO you are a TKL dev! & you too @l-arnold :smile:)

@l-arnold Optional module installs should be easy IMO. The third conf script that you have essentially does it from what I can see. Although there may be some base config which might need to be modified (your config is built to include that stuff - whereas a minimum base config may not).

**Comment by l-arnold** [comment]
@JedMeister - From a fast hack point of view, so long as we can keep an "elaborate" and a "base" config generally aiming for the same folders and config setups my thought is that one way to do a "quick" jump to full build would be the following:

Install Base to Cloud.
Install "full build" to VM.  Run TKLBAM backup
Restore TKLBAM to Cloud and "voila".

That said, every folder, every configuration etc needs to be kept consistent.  Needs to be consistent for V 15 etc for updates.

I am a little worrried about this Repository drifting apart from Kens and the above not being funcitonal.

**Comment by l-arnold** [comment]
btw, thanks for the smile and the "dev" moniker. :+1: :laughing:

**Comment by DocCyblade** [comment]
@l-arnold - My goal is two parts. Part one, make an entry level Odoo TK Build, Part two make it so that it can be customized to the end users liking after install. Part one, should get Odoo in the official repo, Part two is making it even better by allowing post install additions.

I agree that whatever is done needs to stay that way. I would prefer TKL mirror that of install docs from the core dev at Odoo land. This way documentation from Odoo community would make sense in a TKL deployment as well. For me the whole TurnKey motto is, anyone that could download and ISO, install it can have a basic (fill in the blank) open source software. So that it's not confusing, point to the open softwares documentation and config files, etc are all similar. @JedMeister correct me if I am way off the reservation so to speak. 

My goal is to take the current code base, take it down to the frame, the start adding parts back on part by part, and testing as we go along, until we have an understanding of how it all works under the hood so we can later add to it. I plan on using as much if not all the code that is already written. Why re-invent the wheel? Instead, I am making a stock car :car:  by taking apart a bus :bus: , then putting it back together with the same parts and some new paint and creating limo :blue_car:  :smile: Could not find a limo icon!

**Comment by JedMeister** [comment]
@DocCyblade - totaly on the money IMO! :+1: 

FWIW here is the post that was already sitting in my browser...! :smile: 

That should work @l-arnold; although my preference would be for a script that could add the additional modules (or even better a group of scripts that users can manually run). 

If you keep in mind that the conf scripts in a TKLDev appliance build code repo are basically just install scripts that could be run on any Debian PC/server... Running those scripts should install the given software (in conjunction with the files that have been overlayed). In fact if you wanted to you should be able to copy across the overlay files and execute the conf script on Core and you should basically end up with the appliance. 

So moving the conf script for additional modules (conf.d/50_external_modules) out of the build code and into an optional post install script (e.g. `git mv conf.d/50_external_modules overlay/root/install-extras`) is much of the work done! Pulling apart what should be installed by default (to make the MVP) and what should be separate "optional extras" is going to be most of the work. Implementing it should be fairly straight forward. The only other bit will be config changes that are required between MVP and "additional modules".

Then once we have confconsole extended so that it supports additional config then they can be really easily added. Although as we are not clear at this stage what format the extended confconsole may take, so it's probably best to just make it a (some) simple bash script(s)...

**Comment by l-arnold** [comment]
Keep me up on the testing and we should be good.  I am quite skilled at stress tests. :smile

**Comment by l-arnold** [comment]
Been doing a series of migrations.  Brought working system up to dev-wip1 (beta 2).  Thows over well etc.

I would like to find out and document how to apply my SSL cert also to Webmin and Webshell.  They don't seem to fall into my Apache way of doing things.

Will look again today at the DEV branch to see why Openerp Service is not loading properly.

**Comment by DocCyblade** [comment]
Hey guys Just checking in. Been very busy here. I should be back in the game Friday night. I should have the base install compete, then start working on the next. Goal would be to make sure it works as designed.

**Comment by JedMeister** [comment]
@l-arnold 

> I would like to find out and document how to apply my SSL cert also to Webmin and Webshell. They don't seem to fall into my Apache way of doing things.

That's because they don't run on/behind Apache... They both run on their own (mini)servers although as of v14.0 they are both hidden behind (essentially proxied by) stunnel. So you'll need to do some reading (and probably trial and error) on how to configure stunnel to use a 3rd party CA cert. It would be quite good to have documented though!

@DocCyblade - good work mate. I'm going to be away for the weekend so I'll be back on board Sunday night/Monday morning my time...

**Comment by l-arnold** [comment]
@JedMeister  and @DocCyblade 
Just a good heads up.  I just got the XEN Build to install in Linode (2 gb x 40 gb).
Took a while to figure out but in the end it was pretty easy.  
Psyched in that I learned something new!  Will see how it goes from here.

**Comment by l-arnold** [comment]
Closing as has been moved to TKL
